File: astrid/src/androidTest/java/org/tasks/Freeze.java
package org.tasks;

import org.joda.time.DateTime;
import org.joda.time.DateTimeUtils;

import java.util.Date;

import static org.tasks.date.DateTimeUtils.currentTimeMillis;

public class Freeze {

    public static Freeze freezeClock() {
        return freezeAt(currentTimeMillis());
    }

    public static Freeze freezeAt(Date date) {
        return freezeAt(date.getTime());
    }

    public static Freeze freezeAt(DateTime dateTime) {
        return freezeAt(dateTime.getMillis());
    }

    public static Freeze freezeAt(long millis) {
        DateTimeUtils.setCurrentMillisFixed(millis);
        return new Freeze();
    }

    public static void thaw() {
        DateTimeUtils.setCurrentMillisSystem();
    }

    public void thawAfter(@SuppressWarnings("UnusedParameters") Snippet snippet) {
        thaw();
    }
}

File: app/src/test/java/org/tasks/Freeze.kt
package org.tasks

import org.tasks.time.DateTime
import org.tasks.time.DateTimeUtils2.currentTimeMillis
import org.tasks.time.DateTimeUtils2.setCurrentMillisFixed
import org.tasks.time.DateTimeUtils2.setCurrentMillisSystem

class Freeze {
    fun <T> thawAfter(run: () -> T): T {
        try {
            return run()
        } finally {
            thaw()
        }
    }

    companion object {

        fun freezeClock(run: () -> Unit): Any {
            return freezeAt(currentTimeMillis()).thawAfter(run)
        }

        fun <T> freezeAt(dateTime: DateTime, run: () -> T): T {
            return freezeAt(dateTime.millis, run)
        }

        fun <T> freezeAt(timestamp: Long, run: () -> T): T {
            return freezeAt(timestamp).thawAfter(run)
        }

        fun freezeAt(dateTime: DateTime): Freeze {
            return freezeAt(dateTime.millis)
        }

        fun freezeAt(millis: Long): Freeze {
            setCurrentMillisFixed(millis)
            return Freeze()
        }

        fun thaw() {
            setCurrentMillisSystem()
        }
    }
}